elastalert/elastalert.py

```diff
@@ -978,6 +982,33 @@ def get_top_counts(self, rule, starttime, endtime, keys, number=5, qk=None):
             all_counts['top_events_%s' % (key)] = dict(counts[:number])
         return all_counts
 
+    def next_alert_time(self, rule, name, timestamp):
+        """ Calculate an 'until' time and exponent based on how much past the last 'until' we are. """
+        if name in self.silence_cache:
+            last_until, exponent = self.silence_cache[name]
+        else:
+            # If for some reason this isn't cached, pretend it was a week ago
+            last_until = timestamp - datetime.timedelta(weeks=1)
+            exponent = 0
+
+        if not rule.get('exponential_realert'):
+            return timestamp + rule['realert'], 0
+
+        diff = seconds(timestamp - last_until)
+        # If it's been less than twice the alert time, increase exponent
+        if diff < seconds(rule['realert']) * 2 ** exponent:
+            exponent += 1
+        else:
+            # If it's been longer, recursively decrease exponent and compare the remaining time difference
+            while diff > seconds(rule['realert']) * 2 ** exponent and exponent > 0:
```

Comment:
This is acceptable, but I believe you're doing a projection (http://en.wikipedia.org/wiki/Projection_%28linear_algebra%29) iteratively here. You can knock this down from O(n) to O(1).

In practice, the diff will be relatively small, so this won't consume many cycles. But it's worth mentioning here.